text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+
+There are different situations where you would want to use an explicit associated type:
+
+1. The type is central to the trait and deserves to be named.
+1. You want to give users the ability to use concrete types without `#[refine]`.
+1. You want to give generic users of your trait the ability specify a particular type, instead of just bounding it.
+1. You want to give users the ability to easily name and bound the type without using (to-be-RFC'd) special syntax to name the type.
+1. You want the trait to work with dynamic dispatch today.
+1. In the future, you want the associated type to be specified as part of `dyn Trait`, instead of using dynamic dispatch itself.
+
+Using our hypothetical `NewIntoIterator` example, most of these are not met for the `IntoIter` type:
+
+1. While the `Item` type is pretty central to users of the trait, the specific iterator type `IntoIter` is usually not.
+1. The concrete type of an impl may or may not be useful, but usually what's important is the specific extra bounds like `ExactSizeIterator` that a user can use. Using `#[refine]` to explicitly choose to expose this (or a fully concrete type) is not overly burdensome.
+1. Rarely does a function taking `impl IntoIterator` specify a particular iterator type; it would be rare to see a function like this, for example:
+   ```rust
+   fn iterate_over_anything_as_long_as_it_is_vec<T>(
+       vec: impl IntoIterator<IntoIter = std::vec::IntoIter<T>, Item = T>
+   )
+   ```
+1. Bounding the iterator by adding extra bounds like `DoubleEndedIterator` is useful, but not the common case for `IntoIterator`. It therefore shouldn't be overly burdensome to use a (reasonably ergonomic) special syntax in the cases where it's needed.
+1. Using `IntoIterator` with dynamic dispatch would be surprising; more common would be to call `.into_iter()` using static dispatch and then pass the resulting iterator to a function that uses dynamic dispatch.
+1. If we did use `IntoIterator` with dynamic dispatch, the resulting iterator being dynamically dispatched would make the most sense.
+
+Therefore, if we were writing `IntoIterator` today, it would probably use `-> impl Trait` in return position instead of having an explicit `IntoIter` type.
+
+The same is not true for `Iterator::Item`: because `Item` is so central to what an `Iterator` is, and because it rarely makes sense to use an opaque type for the item, it would remain an explicit associated type.
+
+# Prior art
+[prior-art]: #prior-art
+
+There are a number of crates that do desugaring like this manually or with procedural macros. One notable example is [real-async-trait](https://crates.io/crates/real-async-trait).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- None.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+### Naming return types
+
+We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.
```

Comment:
```suggestion
We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.
```

This sentences is ambiguous. At first I thought you meant "introducing a named associated type into the trait", but now I think you mean something like RTN? It'd be useful to reword it for clarity. In particular, I think what this means is "some way for generic code using a trait to place bounds on the return type of `-> impl Trait` functions"? (This is slightly different from being able to name the results of arbitrary `-> impl Trait` functions, e.g., if a function returns `-> (impl Trait, impl Trait)`.)